pilot/cmd/pilot-agent/readiness.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+
+	"errors"
+
+	"time"
+
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	statsRequestTimeout     = time.Second * 2
+	readyStalenessThreshold = time.Minute * 5
+	statXdsActive           = "cluster.xds-grpc.upstream_cx_active"
+	statCdsUpdates          = "cluster_manager.cds.update_success"
+	statWarmingClusters     = "cluster_manager.warming_clusters"
+	statLdsUpdates          = "listener_manager.lds.update_success"
+	statWarmingListeners    = "listener_manager.total_listeners_warming"
+)
+
+type readinessProbe struct {
+	adminPort         int
+	lastReadySnapshot *statsSnapshot
+	lastErr           error
+	mutex             sync.Mutex
+}
+
+func newReadinessProbe(adminPort int) *readinessProbe {
+	return &readinessProbe{
+		adminPort: adminPort,
+	}
+}
+
+func (p *readinessProbe) checkReady() error {
+	// Get a new snapshot of the Envoy stats.
+	s, err := p.newStatsSnapshot()
+
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if err != nil {
+		// Failed getting the current snapshot. Envoy may be restarting. Reinitialize the ready probe.
+		p.lastReadySnapshot = nil
+		p.lastErr = nil
+		return err
+	}
+
+	// Case 1: Wait for the first ready snapshot
+	if p.lastReadySnapshot == nil {
+		if s.active && s.isReady() {
+			// Only save the first ready snapshot
+			return p.transitionToReady(s)
+		}
+		return p.transitionToNotReady(s)
+	}
+
+	// Case 2: the last result was not ready
+	if p.lastErr != nil {
+		if s.active {
+			if s.isReady() && s.isUpdatedSince(p.lastReadySnapshot) {
```

Comment:
Similarly, if proxy received updates since the last ready snapshot, but is not currently connected to Pilot, it should still transition to the ready state.

Reply:
I'm not sure it matters ... if active is flipping on and off that frequently, it's generally in a bad state and I don't think it's that bad that we await going active again.